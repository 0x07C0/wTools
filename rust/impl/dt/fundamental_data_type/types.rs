/// Internal namespace.
mod internal
{
  use crate::exposed::*;

  // xxx : module type_constructors
  // xxx : samples
  // xxx : redo implements

  // xxx : implement clone_as_tuple()
  // xxx : implement clone_as_array()
  // xxx : implement as_tuple()
  // xxx : implement as_array()
  // xxx : implement as_slice()

  // xxx : from type for homopair
  // xxx : from array for homopair
  // xxx : from slice for homopair

  ///
  /// Type constructor to define tuple wrapping a given type.
  ///
  /// Quite often you need to wrap a given type into new one.
  /// For example if orphan rule became and obstacle one should introduce a new type wrapping foreing one.
  /// Type constructr `types!` does exaclty that and auto-implement traits From, Into and Deref for the constructed type.
  ///
  /// ### Sample :: single line single.
  ///
  /// To define your own single use macro `types!`. Single-line definition looks like that.
  ///
  /// ```rust
  /// use fundamental_data_type::prelude::*;
  /// types!( single MySingle : i32 );
  /// let x = MySingle( 13 );
  /// println!( "x : {}", x.0 );
  /// ```
  ///
  /// It gererates code:
  ///
  /// ```rust
  /// use fundamental_data_type::prelude::*;
  ///
  /// pub struct MySingle( pub i32 );
  ///
  /// impl core::ops::Deref for MySingle
  /// {
  ///   type Target = i32;
  ///   fn deref( &self ) -> &Self::Target
  ///   {
  ///     &self.0
  ///   }
  /// }
  /// impl From< i32 > for MySingle
  /// {
  ///   fn from( src : i32 ) -> Self
  ///   {
  ///     Self( src )
  ///   }
  /// }
  /// impl From< MySingle > for i32
  /// {
  ///   fn from( src : MySingle ) -> Self
  ///   {
  ///     src.0
  ///   }
  /// }
  ///
  /// let x = MySingle( 13 );
  /// println!( "x : {}", x.0 );
  /// ```
  ///
  /// ### Sample :: single with derives and attributes.
  ///
  /// It's possible to define attributes as well as derives.
  ///
  /// ```rust
  /// use fundamental_data_type::prelude::*;
  /// types!
  /// {
  ///   /// This is also attribute and macro understands it.
  ///   #[ derive( Debug ) ]
  ///   single MySingle : i32;
  /// }
  /// let x = MySingle( 13 );
  /// dbg!( x );
  /// ```
  ///
  /// It gererates code:
  ///
  /// ```rust
  /// use fundamental_data_type::prelude::*;
  ///
  /// /// This is also attribute and macro understands it.
  /// #[ derive( Debug ) ]
  /// pub struct MySingle( pub i32 );
  ///
  /// impl core::ops::Deref for MySingle
  /// {
  ///   type Target = i32;
  ///   fn deref( &self ) -> &Self::Target
  ///   {
  ///     &self.0
  ///   }
  /// }
  /// impl From< i32 > for MySingle
  /// {
  ///   fn from( src : i32 ) -> Self
  ///   {
  ///     Self( src )
  ///   }
  /// }
  /// impl From< MySingle > for i32
  /// {
  ///   fn from( src : MySingle ) -> Self
  ///   {
  ///     src.0
  ///   }
  /// }
  ///
  /// let x = MySingle( 13 );
  /// dbg!( x );
  /// ```
  ///
  /// ### Sample :: single with struct instead of macro.
  ///
  /// Sometimes it's sufficient to use common type instead of defining a brand new.
  /// You may use paramtetrized struct `Single< T >` instead of macro `types!` if that is the case.
  ///
  /// ```rust
  /// use fundamental_data_type::prelude::*;
  /// let x = Single::< i32 >( 13 );
  /// dbg!( x );
  /// ```
  ///
  /// ### Sample :: single with parametrized element.
  ///
  /// Element of tuple could be parametrized.
  ///
  /// ```rust
  /// use fundamental_data_type::prelude::*;
  /// types!
  /// {
  ///   #[ derive( Debug ) ]
  ///   single MySingle : std::sync::Arc< T : Copy >;
  /// }
  /// let x = MySingle( std::sync::Arc::new( 13 ) );
  /// dbg!( x );
  /// ```
  ///
  /// It gererates code:
  ///
  /// ```rust
  /// use fundamental_data_type::*;
  ///
  /// #[ derive( Debug ) ]
  /// pub struct MySingle< T : Copy >( pub std::sync::Arc< T > );
  ///
  /// impl<T: Copy> core::ops::Deref for MySingle< T >
  /// {
  ///   type Target = std::sync::Arc< T >;
  ///   fn deref( &self ) -> &Self::Target
  ///   {
  ///     &self.0
  ///   }
  /// }
  /// impl< T : Copy > From< std::sync::Arc< T > > for MySingle< T >
  /// {
  ///   fn from( src : std::sync::Arc<T>) -> Self {
  ///     Self( src )
  ///   }
  /// }
  /// impl< T : Copy > From< MySingle< T > > for std::sync::Arc< T >
  /// {
  ///   fn from(src: MySingle<T>) -> Self
  ///   {
  ///     src.0
  ///   }
  /// }
  ///
  /// let x = MySingle( std::sync::Arc::new( 13 ) );
  /// ```
  ///
  /// ### Sample :: single with parametrized tuple.
  ///
  /// Instead of parametrizing the element it's possible to define a parametrized tuple.
  ///
  ///
  /// ```rust
  /// use fundamental_data_type::prelude::*;
  /// types!
  /// {
  ///   #[ derive( Debug ) ]
  ///   single MySingle : < T : Copy >;
  /// }
  /// let x = MySingle( 13 );
  /// dbg!( x );
  /// ```
  ///
  /// It gererates code:
  ///
  /// ```rust
  /// #[ derive( Debug ) ]
  /// pub struct MySingle< T : Copy >( pub T );
  ///
  /// impl< T : Copy > core::ops::Deref
  /// for MySingle< T >
  /// {
  ///   type Target = T;
  ///   fn deref( &self ) -> &Self::Target
  ///   {
  ///     &self.0
  ///   }
  /// }
  ///
  /// impl< T : Copy > From< T >
  /// for MySingle< T >
  /// {
  ///   fn from( src : T ) -> Self
  ///   {
  ///     Self( src )
  ///   }
  /// }
  ///
  /// let x = MySingle( 13 );
  /// dbg!( 13 );
  /// ```
  ///

  // #[ doc = include_str!( concat!( env!( "CARGO_MANIFEST_DIR" ), "/Readme.md" ) ) ]

  #[ macro_export ]
  macro_rules! types
  {

    // No more.

    (
    )
    =>
    {
    };

    // No more.

    (
      ;
    )
    =>
    {
    };

    // single

    (
      $( #[ $Meta : meta ] )*
      single
      $( $Rest : tt )*
    )
    =>
    {
      $crate::_single!
      {
        $( #[ $Meta ] )*
        single
        $( $Rest )*
      }
    };

    // pair

    (
      $( #[ $Meta : meta ] )*
      pair
      $( $Rest : tt )*
    )
    =>
    {
      $crate::_pair!
      {
        $( #[ $Meta ] )*
        pair
        $( $Rest )*
      }
    };

    // bad syntax

    (
      $( $Rest : tt )*
    )
    =>
    {
      compile_error!
      (
        concat!
        (
          "Bad syntax.\n",
          "Expects : {kind} {name} : {type}.\n",
          "For example : `single MySingle : std::sync::Arc< T : Copy >`.\n",
          "But got:\n",
          stringify!
          (
            $( $Rest )*
          ),
        )
      );
    };

  }

  pub use types;
}

/// Protected namespace of the module.
pub mod protected
{
  pub use super::orphan::*;
}

pub use protected::*;

/// Orphan namespace of the module.
pub mod orphan
{
  pub use super::exposed::*;
}

/// Exposed namespace of the module.
pub mod exposed
{
  pub use super::prelude::*;
}

pub use exposed::*;

/// Prelude to use: `use wtools::prelude::*`.
pub mod prelude
{
  #[ doc( inline ) ]
  pub use super::internal::
  {
    types,
  };
}
