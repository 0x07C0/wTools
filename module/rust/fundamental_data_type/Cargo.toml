[package]
name = "fundamental_data_type"
version = "0.1.0"
edition = "2021"
authors = [
  "Kostiantyn Wandalen <wandalen@obox.systems>",
  "Dmytro Kryvoruchko <dm.vict.kr@gmail.com>",
]
license = "MIT"
readme = "Readme.md"
documentation = "https://docs.rs/fundamental_data_type"
repository = "https://github.com/Wandalen/wTools/tree/master/module/rust/fundamental_data_type"
homepage = "https://github.com/Wandalen/wTools/tree/master/module/rust/fundamental_data_type"
description = """
Fundamental data types and type constructors, like Single, Pair, Many.
"""
categories = [ "algorithms", "development-tools" ]
keywords = [ "fundamental", "general-purpose" ]

include = [
  "/rust/impl/dt",
  "/Cargo.toml",
  "/Readme.md",
  "/License",
]

[lib]
name = "fundamental_data_type"
path = "rust/impl/dt/fundamental_data_type/lib.rs"

[[test]]
name = "fundamental_data_type_test"
path = "rust/test/dt/fundamental_data_type/lib_test.rs"

# [[example]]
# name = "fundamental_data_type_trivial_sample"
# path = "sample/rust/fundamental_data_type_trivial_sample/src/main.rs"
# <!-- xxx : write please -->

[dependencies]

[dev-dependencies]
test_tools = { version = "~0.1", path = "../../rust/test_tools" }
