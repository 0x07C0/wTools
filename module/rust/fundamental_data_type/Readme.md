# Module :: fundamental_data_type
[![experimental](https://img.shields.io/badge/stability-experimental-orange.svg)](https://github.com/emersion/stability-badges#experimental) [![rust-status](https://github.com/Wandalen/wTools/actions/workflows/ModuleFundamentalDataTypePush.yml/badge.svg)](https://github.com/Wandalen/wTools/actions/workflows/ModuleFundamentalDataTypePush.yml) [![docs.rs](https://img.shields.io/docsrs/fundamental_data_type?color=e3e8f0&logo=docs.rs)](https://docs.rs/fundamental_data_type) [![discord](https://img.shields.io/discord/872391416519737405?color=e3e8f0&logo=discord&logoColor=e3e8f0)](https://discord.gg/JwTG6d2b)

Fundamental data types and type constructors, like Single, Pair, Many.

# Sample :: trivial

```rust
```
<!-- zzz : write please -->

### To add to your project

``` shell
cargo add fundamental_data_type
```

### Try out from the repository

``` shell test
git clone https://github.com/Wandalen/wTools
cd wTools
cd sample/rust/fundamental_data_type_trivial_sample
cargo run
```


