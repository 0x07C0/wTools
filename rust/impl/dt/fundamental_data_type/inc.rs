
/// Type constructor of pair.
pub mod pair;
/// Type constructor of single.
pub mod single;
/// Type constructors.
pub mod types;
/// Generic traits.
pub mod traits;

/// Protected namespace of the module.
pub mod protected
{
  pub use super::orphan::*;
  pub use super::pair::orphan::*;
  pub use super::single::orphan::*;
  pub use super::types::orphan::*;
  pub use super::traits::orphan::*;
}

pub use protected::*;

/// Orphan namespace of the module.
pub mod orphan
{
  pub use super::exposed::*;
}

/// Exposed namespace of the module.
pub mod exposed
{
  pub use super::prelude::*;
  pub use super::pair::exposed::*;
  pub use super::single::exposed::*;
  pub use super::types::exposed::*;
  pub use super::traits::exposed::*;
}

/// Prelude to use: `use wtools::prelude::*`.
pub mod prelude
{
  pub use super::pair::prelude::*;
  pub use super::single::prelude::*;
  pub use super::types::prelude::*;
  pub use super::traits::prelude::*;
}
