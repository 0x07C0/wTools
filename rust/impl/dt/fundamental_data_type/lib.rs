#![ warn( rust_2018_idioms ) ]
#![ warn( missing_debug_implementations ) ]
#![ warn( missing_docs ) ]

//!
//! Fundamental data types and type constructors, like Single, Pair, Many.
//!

#![ doc = include_str!( concat!( env!( "CARGO_MANIFEST_DIR" ), "/Readme.md" ) ) ]

#[ path = "./inc.rs" ]
mod inc;
pub use inc::*;

// xxx : implement module include_md
