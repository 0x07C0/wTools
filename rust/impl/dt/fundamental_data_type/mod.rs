
#[ path = "./inc.rs" ]
mod inc;
pub use inc::*;
