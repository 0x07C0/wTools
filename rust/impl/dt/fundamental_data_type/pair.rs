// /// Internal namespace.
// mod internal
// {
//   use crate::exposed::*;
//
//   ///
//   /// Type constructor to define tuple wrapping a given type.
//   ///
//   /// Quite often you need to wrap a given type into new one.
//   /// For example if orphan rule became and obstacle one should introduce a new type wrapping foreing one.
//   /// Type constructr `pair!` does exaclty that and auto-implement traits From, Into and Deref for the constructed type.
//   ///
//   /// ### Sample :: pair line pair.
//   ///
//   /// To define your own pair use macro `pair!`. Pair-line definition looks like that.
//   ///
//   /// ```rust
//   /// use fundamental_data_type::prelude::*;
//   /// pair!( MyPair : i32 );
//   /// let x = MyPair( 13 );
//   /// println!( "x : {}", x.0 );
//   /// ```
//   ///
//   /// It gererates code:
//   ///
//   /// ```rust
//   /// use fundamental_data_type::prelude::*;
//   ///
//   /// pub struct MyPair( pub i32 );
//   ///
//   /// impl core::ops::Deref for MyPair
//   /// {
//   ///   type Target = i32;
//   ///   fn deref( &self ) -> &Self::Target
//   ///   {
//   ///     &self.0
//   ///   }
//   /// }
//   /// impl From< i32 > for MyPair
//   /// {
//   ///   fn from( src : i32 ) -> Self
//   ///   {
//   ///     Self( src )
//   ///   }
//   /// }
//   /// impl From< MyPair > for i32
//   /// {
//   ///   fn from( src : MyPair ) -> Self
//   ///   {
//   ///     src.0
//   ///   }
//   /// }
//   ///
//   /// let x = MyPair( 13 );
//   /// println!( "x : {}", x.0 );
//   /// ```
//   ///
//   /// ### Sample :: pair with derives and attributes.
//   ///
//   /// It's possible to define attributes as well as derives.
//   ///
//   /// ```rust
//   /// use fundamental_data_type::prelude::*;
//   /// pair!
//   /// {
//   ///   /// This is also attribute and macro understands it.
//   ///   #[ derive( Debug ) ]
//   ///   MyPair : i32;
//   /// }
//   /// let x = MyPair( 13 );
//   /// dbg!( x );
//   /// ```
//   ///
//   /// It gererates code:
//   ///
//   /// ```rust
//   /// use fundamental_data_type::prelude::*;
//   ///
//   /// /// This is also attribute and macro understands it.
//   /// #[ derive( Debug ) ]
//   /// pub struct MyPair( pub i32 );
//   ///
//   /// impl core::ops::Deref for MyPair
//   /// {
//   ///   type Target = i32;
//   ///   fn deref( &self ) -> &Self::Target
//   ///   {
//   ///     &self.0
//   ///   }
//   /// }
//   /// impl From< i32 > for MyPair
//   /// {
//   ///   fn from( src : i32 ) -> Self
//   ///   {
//   ///     Self( src )
//   ///   }
//   /// }
//   /// impl From< MyPair > for i32
//   /// {
//   ///   fn from( src : MyPair ) -> Self
//   ///   {
//   ///     src.0
//   ///   }
//   /// }
//   ///
//   /// let x = MyPair( 13 );
//   /// dbg!( x );
//   /// ```
//   ///
//   /// ### Sample :: pair with struct instead of macro.
//   ///
//   /// Sometimes it's sufficient to use common type instead of defining a brand new.
//   /// You may use paramtetrized struct `Pair< T >` instead of macro `pair!` if that is the case.
//   ///
//   /// ```rust
//   /// use fundamental_data_type::prelude::*;
//   /// let x = Pair::< i32 >( 13 );
//   /// dbg!( x );
//   /// ```
//   ///
//   /// ### Sample :: pair with parametrized element.
//   ///
//   /// Element of tuple could be parametrized.
//   ///
//   /// ```rust
//   /// use fundamental_data_type::prelude::*;
//   /// pair!
//   /// {
//   ///   #[ derive( Debug ) ]
//   ///   MyPair : std::sync::Arc< T : Copy >;
//   /// }
//   /// let x = MyPair( std::sync::Arc::new( 13 ) );
//   /// dbg!( x );
//   /// ```
//   ///
//   /// It gererates code:
//   ///
//   /// ```rust
//   /// use fundamental_data_type::*;
//   ///
//   /// #[ derive( Debug ) ]
//   /// pub struct MyPair< T : Copy >( pub std::sync::Arc< T > );
//   ///
//   /// impl<T: Copy> core::ops::Deref for MyPair< T >
//   /// {
//   ///   type Target = std::sync::Arc< T >;
//   ///   fn deref( &self ) -> &Self::Target
//   ///   {
//   ///     &self.0
//   ///   }
//   /// }
//   /// impl< T : Copy > From< std::sync::Arc< T > > for MyPair< T >
//   /// {
//   ///   fn from( src : std::sync::Arc<T>) -> Self {
//   ///     Self( src )
//   ///   }
//   /// }
//   /// impl< T : Copy > From< MyPair< T > > for std::sync::Arc< T >
//   /// {
//   ///   fn from(src: MyPair<T>) -> Self
//   ///   {
//   ///     src.0
//   ///   }
//   /// }
//   ///
//   /// let x = MyPair( std::sync::Arc::new( 13 ) );
//   /// ```
//   ///
//   /// ### Sample :: pair with parametrized tuple.
//   ///
//   /// Instead of parametrizing the element it's possible to define a parametrized tuple.
//   ///
//   ///
//   /// ```rust
//   /// use fundamental_data_type::prelude::*;
//   /// pair!
//   /// {
//   ///   #[ derive( Debug ) ]
//   ///   MyPair : < T : Copy >;
//   /// }
//   /// let x = MyPair( 13 );
//   /// dbg!( x );
//   /// ```
//   ///
//   /// It gererates code:
//   ///
//   /// ```rust
//   /// #[ derive( Debug ) ]
//   /// pub struct MyPair( pub i32 );
//   ///
//   /// impl core::ops::Deref for MyPair
//   /// {
//   ///   type Target = i32;
//   ///   fn deref( &self ) -> &Self::Target
//   ///   {
//   ///     &self.0
//   ///   }
//   /// }
//   /// impl From< i32 > for MyPair
//   /// {
//   ///   fn from( src : i32 ) -> Self
//   ///   {
//   ///     Self( src )
//   ///   }
//   /// }
//   /// impl From< MyPair > for i32
//   /// {
//   ///   fn from( src : MyPair ) -> Self
//   ///   {
//   ///     src.0
//   ///   }
//   /// }
//   /// let x = MyPair( 13 );
//   /// dbg!( 13 );
//   /// ```
//   ///
//
//   // #[ doc = include_str!( concat!( env!( "CARGO_MANIFEST_DIR" ), "/Readme.md" ) ) ]
//
//   #[ macro_export ]
//   macro_rules! pair
//   {
//
//     // Pair : < T >;
//
//     (
//       $( #[ $Meta : meta ] )*
//       $Name : ident :
//       < $ParamName : ident $( : $ParamTy1x1 : ident $( :: $ParamTy1xN : ident )* $( + $ParamTy2 : path )* )? >
//       $(;)?
//     )
//     =>
//     {
//       $( #[ $Meta ] )*
//       pub struct $Name
//       < $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? >
//       ( pub $ParamName );
//
//       impl< $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? > core::ops::Deref
//       for $Name
//       < $ParamName >
//       {
//         type Target = $ParamName;
//         fn deref( &self ) -> &Self::Target
//         {
//           &self.0
//         }
//       }
//
//       impl< $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? > From< $ParamName >
//       for $Name
//       < $ParamName >
//       {
//         fn from( src : $ParamName ) -> Self
//         {
//           Self( src )
//         }
//       }
//
//       // From Pair Into Element cant be implemented because of Rust restructions.
//
//     };
//
//     // Pair : < T1, ... >;
//
//     (
//       $( #[ $Meta : meta ] )*
//       $Name : ident :
//       < $ParamName : ident $( : $ParamTy1x1 : ident $( :: $ParamTy1xN : ident )* $( + $ParamTy2 : path )* )? ,
//       $( $Rest : tt )*
//     )
//     =>
//     {
//       compile_error!
//       (
//         concat!
//         (
//           "Parametrized element should be pair, because Pair has only one element\n",
//           stringify!
//           (
//             $( #[ $Meta ] )*
//             $Name :
//             < $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? ,
//             $( $Rest )*
//           )
//         )
//       );
//     };
//
//     // Pair : Element< T1, T2, ... >;
//
//     (
//       $( #[ $Meta : meta ] )*
//       $Name : ident : $TypeSplit1 : ident $( :: $TypeSplitN : ident )*
//       < $( $ParamName : ident $( : $ParamTy1x1 : ident $( :: $ParamTy1xN : ident )* $( + $ParamTy2 : path )* )? ),* >
//       $(;)?
//     )
//     =>
//     {
//       $( #[ $Meta ] )*
//       pub struct $Name
//       < $( $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? ),* >
//       ( pub $TypeSplit1 $( :: $TypeSplitN )* < $( $ParamName ),* > );
//
//       impl< $( $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? ),* > core::ops::Deref
//       for $Name
//       < $( $ParamName ),* >
//       {
//         type Target = $TypeSplit1 $( :: $TypeSplitN )* < $( $ParamName ),* >;
//         fn deref( &self ) -> &Self::Target
//         {
//           &self.0
//         }
//       }
//
//       impl< $( $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? ),* >
//       From< $TypeSplit1 $( :: $TypeSplitN )* < $( $ParamName ),* > >
//       for $Name
//       < $( $ParamName ),* >
//       {
//         fn from( src : $TypeSplit1 $( :: $TypeSplitN )* < $( $ParamName ),* > ) -> Self
//         {
//           Self( src )
//         }
//       }
//
//       impl< $( $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? ),* >
//       From< $Name< $( $ParamName ),* > >
//       for $TypeSplit1 $( :: $TypeSplitN )* < $( $ParamName ),* >
//       {
//         fn from( src : $Name< $( $ParamName ),* > ) -> Self
//         {
//           src.0
//         }
//       }
//
//     };
//
//     // Pair : Element;
//
//     (
//       $( #[ $Meta : meta ] )*
//       $Name : ident : $TypeSplit1 : ident $( :: $TypeSplitN : ident )* $(;)?
//       // $Name : ident : $Type : ty $(;)?
//     )
//     =>
//     {
//       $crate::pair!
//       (
//         $( #[ $Meta ] )*
//         $Name : $TypeSplit1 $( :: $TypeSplitN )* <>;
//       );
//
//     };
//
//   }
//
// //   pair!
// //   {
// //
// //     ///
// //     /// A type wrapping a another type into a tuple.
// //     ///
// //     ///
// //     /// ### Sample :: struct instead of macro.
// //     ///
// //     /// Sometimes it's sufficient to use common type instead of defining a brand new one.
// //     /// You may use paramtetrized struct `fundamental_data_type::Pair< T >` instead of macro `fundamental_data_type::pair!` if that is the case.
// //     ///
// //     /// ```rust
// //     /// use fundamental_data_type::prelude::*;
// //     /// let x = Pair::< i32 >( 13 );
// //     /// dbg!( x );
// //     /// ```
// //     ///
// //
// //     #[ derive( Debug, Clone, PartialEq, Eq ) ]
// //     Pair : < T >;
// //
// //   }
//
//   pub use pair;
// }

/// Protected namespace of the module.
pub mod protected
{
  pub use super::orphan::*;
}

pub use protected::*;

/// Orphan namespace of the module.
pub mod orphan
{
  pub use super::exposed::*;
}

/// Exposed namespace of the module.
pub mod exposed
{
  pub use super::prelude::*;
}

pub use exposed::*;

/// Prelude to use: `use wtools::prelude::*`.
pub mod prelude
{
  // pub use super::internal::
  // {
  //   // pair,
  //   // Pair,
  // };
}
