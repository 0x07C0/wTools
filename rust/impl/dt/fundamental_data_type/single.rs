/// Internal namespace.
mod internal
{
  use crate::exposed::*;

  // xxx : check doc

  ///
  /// Type constructor of single.
  ///
  /// Should not be used directly. Instead use macro [crate::type!].
  ///

  #[ macro_export ]
  macro_rules! _single
  {

    // single Single : < T >;

    (
      $( #[ $Meta : meta ] )*
      single $Name : ident :
      < $ParamName : ident $( : $ParamTy1x1 : ident $( :: $ParamTy1xN : ident )* $( + $ParamTy2 : path )* )? >
      $( ; $( $Rest : tt )* )?
    )
    =>
    {
      $( #[ $Meta ] )*
      pub struct $Name
      < $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? >
      ( pub $ParamName );

      impl< $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? > core::ops::Deref
      for $Name
      < $ParamName >
      {
        type Target = $ParamName;
        fn deref( &self ) -> &Self::Target
        {
          &self.0
        }
      }

      impl< $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? > From< $ParamName >
      for $Name
      < $ParamName >
      {
        fn from( src : $ParamName ) -> Self
        {
          Self( src )
        }
      }

      // From Single Into Element cant be implemented because of Rust restructions.

      $crate::types!{ $( $( $Rest )* )? }
    };

    // single Single : < T1, ... >;

    (
      $( #[ $Meta : meta ] )*
      single $Name : ident :
      < $ParamName : ident $( : $ParamTy1x1 : ident $( :: $ParamTy1xN : ident )* $( + $ParamTy2 : path )* )? ,
      $( $Rest : tt )*
    )
    =>
    {
      compile_error!
      (
        concat!
        (
          "Parametrized element should be single, because Single has only one element\n",
          stringify!
          (
            $( #[ $Meta ] )*
            $Name :
            < $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? ,
            $( $Rest )*
          )
        )
      );
    };

    // single Single : Element< T1, T2, ... >;

    (
      $( #[ $Meta : meta ] )*
      single $Name : ident : $TypeSplit1 : ident $( :: $TypeSplitN : ident )*
      < $( $ParamName : ident $( : $ParamTy1x1 : ident $( :: $ParamTy1xN : ident )* $( + $ParamTy2 : path )* )? ),* >
      $( ; $( $Rest : tt )* )?
    )
    =>
    {
      $( #[ $Meta ] )*
      pub struct $Name
      < $( $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? ),* >
      ( pub $TypeSplit1 $( :: $TypeSplitN )* < $( $ParamName ),* > );

      impl< $( $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? ),* > core::ops::Deref
      for $Name
      < $( $ParamName ),* >
      {
        type Target = $TypeSplit1 $( :: $TypeSplitN )* < $( $ParamName ),* >;
        fn deref( &self ) -> &Self::Target
        {
          &self.0
        }
      }

      impl< $( $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? ),* >
      From< $TypeSplit1 $( :: $TypeSplitN )* < $( $ParamName ),* > >
      for $Name
      < $( $ParamName ),* >
      {
        fn from( src : $TypeSplit1 $( :: $TypeSplitN )* < $( $ParamName ),* > ) -> Self
        {
          Self( src )
        }
      }

      impl< $( $ParamName $( : $ParamTy1x1 $( :: $ParamTy1xN )* $( + $ParamTy2 )* )? ),* >
      From< $Name< $( $ParamName ),* > >
      for $TypeSplit1 $( :: $TypeSplitN )* < $( $ParamName ),* >
      {
        fn from( src : $Name< $( $ParamName ),* > ) -> Self
        {
          src.0
        }
      }

      $crate::types!{ $( $( $Rest )* )? }
    };

    // single Single : Element;

    (
      $( #[ $Meta : meta ] )*
      single $Name : ident : $TypeSplit1 : ident $( :: $TypeSplitN : ident )*
      $( ; $( $Rest : tt )* )?
    )
    =>
    {
      $crate::types!
      (
        $( #[ $Meta ] )*
        single $Name : $TypeSplit1 $( :: $TypeSplitN )* <>;
        // $( ; $( $Rest )* )?
      );
      $crate::types!{ $( $( $Rest )* )? }
    };

  }

  types!
  {

    ///
    /// Type constructor to wrap a another type into a tuple.
    ///
    /// ### Sample :: struct instead of macro.
    ///
    /// Sometimes it's sufficient to use common type instead of defining a brand new one.
    /// You may use paramtetrized struct `fundamental_data_type::Single< T >` instead of macro `fundamental_data_type::types!` if that is the case.
    ///
    /// ```rust
    /// use fundamental_data_type::prelude::*;
    /// let x = Single::< i32 >( 13 );
    /// dbg!( x );
    /// ```
    ///

    #[ derive( Debug, Clone, PartialEq, Eq, Default ) ]
    single Single : < T >;

  }

  pub use _single;
}

/// Protected namespace of the module.
pub mod protected
{
  pub use super::orphan::*;
}

pub use protected::*;

/// Orphan namespace of the module.
pub mod orphan
{
  pub use super::exposed::*;
}

/// Exposed namespace of the module.
pub mod exposed
{
  pub use super::prelude::*;
}

pub use exposed::*;

/// Prelude to use: `use wtools::prelude::*`.
pub mod prelude
{
  pub use super::internal::
  {

    _single,
    Single,

  };
}