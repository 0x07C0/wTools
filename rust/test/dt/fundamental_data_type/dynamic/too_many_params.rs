use fundamental_data_type as TheModule;
use TheModule::prelude::*;

types!
{
  single Single : < T1, T2 >;
}

fn main()
{
}
