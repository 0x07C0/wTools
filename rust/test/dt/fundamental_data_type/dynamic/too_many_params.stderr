error: Parametrized element should be single, because Single has only one element
       Single : < T1, T2 > ;
 --> ../../../rust/test/dt/fundamental_data_type/dynamic/too_many_params.rs:4:1
  |
4 | / types!
5 | | {
6 | |   single Single : < T1, T2 >;
7 | | }
  | |_^
  |
  = note: this error originates in the macro `$crate::_single` (in Nightly builds, run with -Z macro-backtrace for more info)
